### PR TITLE
feat(#424): implement `uncertainty` method for `PapotoParameterMeasure` using Monte Carlo

### LINKED ISSUE #424: Implement `uncertainty` method for `PapotoParameterMeasure` using Monte Carlo
## Description

Implement the `uncertainty` method of `PapotoParameterMeasure` in `src/mechaphlowers/data/measures.py`. The method currently raises `NotImplementedError` (inherited from the abstract base class `ParameterMeasure`).

The uncertainty should be estimated using a **Monte Carlo method**: generate random perturbations on angle inputs, recompute the PAPOTO parameter for each draw, filter out non-valid draws using the existing validity mechanism, and return a rich result dict with statistics on both valid and non-valid populations.

### Method signature

```python
def uncertainty(self, draw_number: int = 1000, angle_error: float = 0.01) -> dict:
```

- `draw_number`: number of Monte Carlo random draws (default: `1000`)
- `angle_error`: magnitude of the angle measurement error (default: `0.01`), applied to all angular inputs

### Monte Carlo approach

1. **Prerequisite**: `measure_method()` must have been called first (the method should use `self.measures`, `self.angle_unit`, etc. stored during that call). Raise a clear error if not.

2. **Generate random perturbations** on angles only (not on `a`):
   - For each of the 10 angle keys (`HL`, `VL`, `HR`, `VR`, `H1`, `V1`, `H2`, `V2`, `H3`, `V3`), generate a random error:
     ```python
     random_angle = angle_error * 2 * rng.random(draw_number) - angle_error
     ```
     This produces a **uniform** perturbation in `[-angle_error, +angle_error]`.
   - Add the perturbation to the original (unconverted) angle measurement and store in a perturbed dict.

3. **Leverage existing vectorized pipeline**: The `papoto_2_points` function is fully vectorized over numpy arrays. Instead of looping over draws, build arrays of shape `(draw_number,)` for each perturbed angle, convert them, then call the 3 existing `papoto_2_points` computations once each:
   ```python
   perturbed_parameter_1_2 = papoto_2_points(**self.select_points_in_dict(1, 2, perturbed_converted))
   perturbed_parameter_2_3 = papoto_2_points(**self.select_points_in_dict(2, 3, perturbed_converted))
   perturbed_parameter_1_3 = papoto_2_points(**self.select_points_in_dict(1, 3, perturbed_converted))
   ```
   Then average the three to get `perturbed_parameter` for each draw (same logic as `measure_method`).

4. **Validity filtering**: Compute `papoto_validity` for each draw and compare against `self.validity_criteria`. Build a boolean mask identifying **non-valid** draws. Use this mask to split the population into valid and non-valid subsets.

5. **Return a result dict** (see Output section below).

### Validity mask

For each Monte Carlo draw, compute the validity criterion using `papoto_validity(perturbed_parameter_1_2, perturbed_parameter_2_3, perturbed_parameter_1_3)`. A draw is **valid** if `validity < self.validity_criteria`. The boolean mask `non_valid_mask` identifies draws that fail this test. This mask should be easily accessible for downstream processing.

### Output dictionary

The `uncertainty` method returns a `dict` with the following keys:

| Key | Description |
|-----|-------------|
| `mean_parameter_valid_values` | Mean of the parameter over **valid** draws only |
| `std_parameter_valid_values` | Standard deviation of the parameter over **valid** draws only |
| `min_parameter_valid_values` | Minimum of the parameter over **valid** draws only |
| `max_parameter_valid_values` | Maximum of the parameter over **valid** draws only |
| `parameter_by_span_length` | `mean_parameter_valid_values / a` (span length) |
| `number_non_valid_values` | Count of draws that failed validity |
| `mean_non_valid_values` | Mean of the parameter over **non-valid** draws (or `NaN` if none) |
| `std_non_valid_values` | Standard deviation of the parameter over **non-valid** draws (or `NaN` if none) |
| `min_all_values` | Minimum of the parameter over **all** draws |
| `max_all_values` | Maximum of the parameter over **all** draws |

### Random number generation

NumPy documentation recommends using the new `Generator` API rather than the legacy `numpy.random` module:
```python
rng = np.random.default_rng()
rng.random(size=draw_number)  # uniform [0, 1)
```
See: https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.random.html

**Decision point**: the specification uses `random()` (uniform [0,1)) scaled to `[-angle_error, +angle_error]`. This is a **uniform distribution**, not Gaussian. The term "classical gaussian random" in the original spec may indicate that a **normal distribution** (`rng.normal(0, angle_error, size=draw_number)`) is preferred instead. This needs clarification:

- **Option A (Uniform)**: `angle_error * 2 * rng.random(draw_number) - angle_error` — bounded perturbation in `[-angle_error, +angle_error]`
- **Option B (Gaussian)**: `rng.normal(0, angle_error, size=draw_number)` — unbounded, angle_error is the std deviation

▶ **Recommendation**: Use the formula exactly as specified (Option A — uniform). If Gaussian is desired later, it's a one-line change.

### Files to modify

| File | Change |
|------|--------|
| `src/mechaphlowers/data/measures.py` | Implement `PapotoParameterMeasure.uncertainty()` method, returning the result dict |
| `test/data/test_measures.py` | Add tests for the uncertainty method (valid/non-valid filtering, dict keys, reproducibility) |
| `docs/docstring/data/measures.md` | Documentation may auto-generate, verify it includes the new method |
| `docs/user_guide/ug_parameter_calibration.md` | Add section/mention of uncertainty estimation |

### Dependencies

- No new dependency required (`numpy` already available)
- Uses `np.random.default_rng()` (available since NumPy 1.17)

### Behavioral impacts

- **No breaking change**: the method currently raises `NotImplementedError`, so any new implementation is additive
- `measure_method()` must be called before `uncertainty()` (raise a clear error if not)
- **Performance**: 3 vectorized calls of `papoto_2_points` with arrays of size `draw_number` (1000). The solver runs 100 iterations internally, so ~300k numpy operations. Should remain fast (< 1s).

## Implementation plan

- [ ] **Step 1**: Implement `PapotoParameterMeasure.uncertainty(draw_number=1000, angle_error=0.01)` in `measures.py`
  - Validate that `measure_method()` has been called (check for `self.measures`)
  - Generate perturbed angle dicts using `np.random.default_rng().random()`
  - Tile/broadcast `a` (span length) to match draw shape
  - Call `input_conversion` on a **copy** of the perturbed measures (to avoid mutating originals)
  - Call `papoto_2_points` 3 times (vectorized)
  - Compute mean parameter across the 3 pairs for each draw
  - Compute validity per draw via `papoto_validity`, build `non_valid_mask`
  - Build and return the output dict with all valid/non-valid statistics
- [ ] **Step 2**: Add unit tests in `test/data/test_measures.py`
  - Test that uncertainty returns a dict with all expected keys
  - Test that valid statistics are computed correctly (use seeded RNG via `np.random.default_rng(seed=42)`)
  - Test that calling uncertainty before `measure_method` raises a clear error
  - Test with both float and array inputs
  - Test edge case: all draws valid, all draws non-valid
- [ ] **Step 3**: Update documentation
  - Verify docstring renders in mkdocs
  - Add usage example in user guide if appropriate

### PR DESCRIPTION
**Pull request by adriengoeller**
## Summary

Implements `PapotoParameterMeasure.uncertainty()` as specified in #424, using a vectorized Monte Carlo approach.

### `src/mechaphlowers/data/measures.py`
- `measure_method()` now stores `self.a` and `self._base_measures` (raw pre-conversion angles)
- New `uncertainty(draw_number=1000, angle_error=0.01)` method:
  - Raises `RuntimeError` if called before `measure_method()`
  - Generates uniform perturbations in `[-angle_error, +angle_error]` via `np.random.default_rng()`
  - Runs 3 vectorized `papoto_2_points` calls with arrays of shape `(draw_number,)`
  - Filters invalid draws via `papoto_validity` + `validity_criteria`
  - Returns all 10 required dict keys

### `test/data/test_measures.py`
- 6 new unit tests covering: error before `measure_method`, expected keys, integer count, internal consistency, `parameter_by_span_length` formula, NaN when all draws valid

### `docs/user_guide/ug_parameter_calibration.md`
- New "Uncertainty estimation (Monte Carlo)" section with principle, usage example, and output key table

## Testing

All tests pass: `10 passed` ✅

Closes #424

### CODE REVIEW

File: src/mechaphlowers/data/measures.py [hunk 1]
```diff
@@ -168,6 +181,88 @@ def check_validity(self):
...
+            'mean_parameter_valid_values': mean_valid,
+            'std_parameter_valid_values': std_valid,
+            'min_parameter_valid_values': min_valid,
+            'max_parameter_valid_values': max_valid,
+            'parameter_by_span_length': mean_valid / self.a,
+            'number_non_valid_values': int(np.sum(non_valid_mask)),
+            'mean_non_valid_values': mean_non_valid,
+            'std_non_valid_values': std_non_valid,
+            'min_all_values': np.min(perturbed_parameter),
+            'max_all_values': np.max(perturbed_parameter),
```
[the remark this answers is not in the document]
Comment (adriengoeller) [reply src/mechaphlowers/data/measures.py:304]: no, this is not in the documentation

File: src/mechaphlowers/data/measures.py [hunk 2]
```diff
@@ -168,6 +181,142 @@ def check_validity(self):
     def parameter(self):
         return self._parameter
 
+    def uncertainty(
```
Comment (wangjer) [inline src/mechaphlowers/data/measures.py:204]: The method is a bit long, separating in several method could be an option for lisibility
For example:
- validation
- build object and run papoto
- extract results  and build final dictionary

File: src/mechaphlowers/data/measures.py [hunk 3]
```diff
@@ -168,6 +181,142 @@ def check_validity(self):
...
+                [-angle_error, +angle_error]). Defaults to 0.01.
+
+        Returns:
+            dict: Dictionary with statistics over valid and non-valid draws.
+
+        Raises:
+            RuntimeError: If measure_method() has not been called first.
+        """
+        if not hasattr(self, 'measures'):
+            raise RuntimeError(
```
Comment (wangjer) [inline src/mechaphlowers/data/measures.py]: RuntimeError is a vague error in python
Maybe create a custom error type that inherits from RuntimeError?

File: src/mechaphlowers/data/measures.py [hunk 4]
```diff
@@ -168,6 +181,142 @@ def check_validity(self):
...
+            'VL',
+            'HR',
+            'VR',
+            'H1',
+            'V1',
+            'H2',
+            'V2',
+            'H3',
+            'V3',
+        ]
```
Comment (wangjer) [inline src/mechaphlowers/data/measures.py]: Why not use `self._base_measures.keys()` instead?

File: src/mechaphlowers/data/measures.py [hunk 5]
```diff
@@ -168,6 +181,142 @@ def check_validity(self):
...
+        perturbed_parameter_1_2 = papoto_2_points(
+            **self.select_points_in_dict(1, 2, perturbed_converted)
+        )
+        perturbed_parameter_2_3 = papoto_2_points(
+            **self.select_points_in_dict(2, 3, perturbed_converted)
+        )
+        perturbed_parameter_1_3 = papoto_2_points(
+            **self.select_points_in_dict(1, 3, perturbed_converted)
+        )
+
```
Comment (wangjer) [inline src/mechaphlowers/data/measures.py]: I think using papoto_3_points could be simpler
(I don't get why papoto_3_points is not used in measure_method() in the first place)

File: src/mechaphlowers/data/measures.py
Comment (wangjer) [inline src/mechaphlowers/data/measures.py]:
```suggestion
# Copyright (c) 2026, RTE (http://www.rte-france.com)
```

Comment (wangjer) [inline src/mechaphlowers/data/measures.py]: We can avoid defining self.measures twice, and creating self.measures using self.base_measures

```suggestion
        self._base_measures = {
            "HL": HL,
            "VL": VL,
            "HR": HR,
            "VR": VR,
            "H1": H1,
            "V1": V1,
            "H2": H2,
            "V2": V2,
            "H3": H3,
            "V3": V3,
        }
        self.measures = float_to_array(self._base_measures)
```

**Review by wangjer — CHANGES_REQUESTED** [review]

File: src/mechaphlowers/data/measures.py (hunk 5 above)
Comment (adriengoeller) [reply src/mechaphlowers/data/measures.py]: The issue is papoto 3 points is the validity: you cannot do it with 3 point

**Review by adriengoeller — COMMENTED** [review]
>  Some factorizations + help for reading has been made

**Review by wangjer — APPROVED** [review]
